### PR TITLE
Detect invalid (unsupported) compression types when parsing

### LINKED ISSUE #4636: Returns error too late when parsing invalid file compression type.
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
When parsing a SQL query with an invalid file compression type, we detect the error when creating the `ListingTable` which is too late:
```sql
❯ create external table t1 stored as csv with header row compression type zzz location "../datafusion/core/tests/example.csv";
Execution("Unknown FileCompressionType ZZZ")
```
Besides, we should return a `ParserError`, instead of a `DataFusionError::Execution` error.
https://github.com/apache/arrow-datafusion/blob/3611d911a3c9f3740bb1fc0527198be39ff47bfd/datafusion/core/src/datasource/listing_table_factory.rs#L62-L70

**Describe the solution you'd like**
We could detect this error when parsing the `CreateExternalTable`

**Describe alternatives you've considered**
We could not do this.

**Additional context**

### PR DESCRIPTION
**Pull request by HaoYang670**
# Which issue does this PR close?

Closes #4636.

# Rationale for this change

Before the change:
```sql
❯ create external table t1 stored as csv with header row compression type zzz location "../datafusion/core/tests/example.csv";
Execution("Unknown FileCompressionType ZZZ")
```

After the change:
```sql
❯ create external table t1 stored as csv with header row compression type zzz location "../datafusion/core/tests/example.csv";  🤔 Invalid statement: sql parser error: Unsupported file compression type ZZZ
```

# What changes are included in this PR?

# Are these changes tested?
Yes

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/datasource/file_format/file_type.rs [hunk 1]
```diff
@@ -50,48 +52,59 @@ pub trait GetExt {
...
+    /// Bzip2-ed file
+    pub const BZIP2: Self = Self { variant: BZIP2 };
+
+    /// Xz-ed file (liblzma)
+    pub const XZ: Self = Self { variant: XZ };
+
+    /// Uncompressed file
+    pub const UNCOMPRESSED: Self = Self {
+        variant: UNCOMPRESSED,
+    };
```
Comment (HaoYang670) [inline datafusion/core/src/datasource/file_format/file_type.rs:101]: They are a little tricky, but can minimize the code change.

File: datafusion/sql/src/parser.rs [hunk 2]
```diff
@@ -66,7 +63,7 @@ pub struct CreateExternalTable {
     /// Option to not error if table already exists
     pub if_not_exists: bool,
     /// File compression type (GZIP, BZIP2, XZ)
-    pub file_compression_type: String,
+    pub file_compression_type: CompressionTypeVariant,
```
Comment (HaoYang670) [inline datafusion/sql/src/parser.rs:66]: Strength the type can let us find errors earlier.

File: datafusion/expr/src/logical_plan/plan.rs [hunk 3]
```diff
@@ -1487,7 +1488,7 @@ pub struct CreateExternalTable {
     /// SQL used to create the table, if available
     pub definition: Option<String>,
     /// File compression type (GZIP, BZIP2, XZ)
-    pub file_compression_type: String,
+    pub file_compression_type: CompressionTypeVariant,
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:1491]: 👍

File: datafusion/sql/src/parser.rs [hunk 4]
```diff
@@ -724,4 +726,13 @@ mod tests {
 
         Ok(())
     }
+
+    #[test]
+    fn invalid_compression_type() {
+        let sql = "CREATE EXTERNAL TABLE t STORED AS CSV COMPRESSION TYPE ZZZ LOCATION 'blahblah'";
+        expect_parse_error(
+            sql,
+            "sql parser error: Unsupported file compression type ZZZ",
```
Comment (alamb) [inline datafusion/sql/src/parser.rs:735]: 👍

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @HaoYang670 

The clippy CI check has failed on this PR, but I think that is because a new version of Rust was released -- if you merge from main it should be solved I think

**Comment by alamb** [comment]
I took the liberty of merging up from master to resolve another conflict and plan to merge this PR when CI passes

**Comment by HaoYang670** [comment]
Thanks @alamb !

**Pull request merged**